airbyte-featureflag/src/main/kotlin/Client.kt

```diff
@@ -0,0 +1,196 @@
+        }
+    }
+}
+
+/**
+ * Cloud feature-flag client. Feature-flags are derived from an external source (the LDClient).
+ * Also supports flags defined via environment-variables via the [EnvVar] class.
+ *
+ * @param [client] the Launch-Darkly client for interfacing with Launch-Darkly.
+ */
+class CloudClient(private val client: LDClient) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> flag.enabled()
+            else -> client.boolVariation(flag.key, ctx.toLDUser(), flag.default)
+        }
+    }
+}
+
+/**
+ * Test feature-flag client. Intended only for usage in testing scenarios.
+ *
+ * All [Flag] instances will use the provided [values] map as their source of truth, including [EnvVar] flags.
+ *
+ * @param [values] is a map of [Flag.key] to enabled/disabled status.
+ */
+class TestClient(val values: Map<String, Boolean>) : FeatureFlagClient {
+    override fun enabled(flag: Flag, ctx: Context): Boolean {
+        return when (flag) {
+            is EnvVar -> {
+                // convert to a EnvVar flag with a custom fetcher that uses the [values] of this Test class
+                // instead of fetching from the environment variables
+                EnvVar(envVar = flag.key, default = flag.default, team = flag.team) {
+                    values[flag.key]?.toString() ?: flag.default.toString()
+                }.enabled()
+            }
+
+            else -> values[flag.key] ?: flag.default
+        }
+    }
+}
+
+
+/**
+ * Data wrapper around OSS feature-flag configuration file.
+ *
+ * The file has the format of:
+ * flags:
+ *  - name: feature-one
+ *    enabled: true
+ *  - name: feature-two
+ *    enabled: false
+ */
+private data class PlatformFlags(val flags: List<PlatformFlag>)
+
+/**
+ * Data wrapper around an individual flag read from the configuration file.
+ */
+private data class PlatformFlag(val name: String, val enabled: Boolean)
+
+
+/** The yaml mapper is used for reading the feature-flag configuration file. */
+private val yamlMapper = ObjectMapper(YAMLFactory()).registerKotlinModule()
+
+/**
+ * Reads a yaml configuration file, converting it into a map of flag name to flag configuration.
+ *
+ * @param [path] to yaml config file
+ * @return map of feature-flag name to feature-flag config
+ */
+//private fun readConfig(path: Path): Map<String, OSSFlag> = yamlMapper.readValue<List<OSSFlag>>(path.toFile())
+//    .associateBy { it.name }
+private fun readConfig(path: Path): Map<String, PlatformFlag> = yamlMapper.readValue<PlatformFlags>(path.toFile())
+    .let { it.flags }
+    .associateBy { it.name }
+
+/**
+ * Monitors a [Path] for changes, calling [block] when a change is detected.
+ *
+ * @receiver Path
+ * @param [block] function called anytime a change is detected on this [Path]
+ */
+private fun Path.onChange(block: () -> Unit) {
+    val watcher = fileSystem.newWatchService()
```

Comment:
Could we make the type to be explicit when it val comes from a method return, it makes it easier to understand when reviewing. IMO:
val string = "str" is ok 
but
val methodReturn = test()
should be
val methodReturn: Type = test()

This comes from the experience I have with scala, non explicit types can become a pain.

Reply:
Kotlin allows this, but I would prefer to have a discussion about these prior pain-points. Both you and Jimmy have brought this up in this PR (and also asked this same question when I gave the presentation months ago).   

Myself, and my previous teams, having written in kotlin, go, and java (using `var`) without running into issues related to not explicitly including the type.  I'm not against adding it, but I want to learn _why_ and _what_ problem that solves.

In this PR I only included the type (only in tests) when casting a class to the interface it implements:
```kotlin
var client: FeatureFlagClient = PlatformClient(/*...*/)
```

I'll schedule some time next week for us to discuss this.